peekingduck/pipeline/nodes/dabble/camera_calibration.py

```diff
@@ -0,0 +1,533 @@
+        np_mgrid = np.mgrid[0:grid_height, 0:grid_width]
+        self.object_points_base = np.zeros((grid_height * grid_width, 3), np.float32)
+        self.object_points_base[:, :2] = np_mgrid.T.reshape(-1, 2)
+
+        # arrays to store object points and image points
+        self.object_points: List[np.ndarray] = []  # points in real world
+        self.image_points: List[np.ndarray] = []  # points on image plane
+
+        self.last_detection = time.time()
+        self.num_detections = 0
+
+        self.num_pictures = 5
+
+    def _detect_corners(self, height: int, width: int, gray_img: np.ndarray) -> tuple:
+        """Detects corners in the image"""
+        # downscale
+        new_h = int(height / self.scale_factor)
+        new_w = int(width / self.scale_factor)
+
+        resized_img = cv2.resize(gray_img, (new_w, new_h), interpolation=cv2.INTER_AREA)
+
+        # try to find chessboard corners
+        return cv2.findChessboardCorners(
+            image=resized_img, patternSize=self.num_corners, corners=None
+        )
+
+    def _calculate_coeffs(self, img_shape: tuple) -> None:
+        """Performs calculations with detected corners"""
+        (
+            calibration_success,
+            camera_matrix,
+            distortion_coeffs,
+            rotation_vec,
+            translation_vec,
+        ) = cv2.calibrateCamera(
+            objectPoints=self.object_points,
+            imagePoints=self.image_points,
+            imageSize=img_shape,
+            cameraMatrix=None,
+            distCoeffs=None,
+        )
+
+        if calibration_success:
+            self.logger.info("Calibration successful!")
+            self.logger.info(f"camera_matrix: {camera_matrix}")
+            self.logger.info(f"distortion_coeffs: {distortion_coeffs}")
+
+            file_data = {}
+            file_data["camera_matrix"] = camera_matrix.tolist()
+            file_data["distortion_coeffs"] = distortion_coeffs.tolist()
+
+            yaml.dump(file_data, open(self.file_path, "w"), default_flow_style=None)
+        else:
+            raise Exception("Calibration failed. Please try again.")
+
+        mean_error = 0
+        for i in range(len(self.object_points)):
+            projected_image_points, _ = cv2.projectPoints(
+                objectPoints=self.object_points[i],
+                rvec=rotation_vec[i],
+                tvec=translation_vec[i],
+                cameraMatrix=camera_matrix,
+                distCoeffs=distortion_coeffs,
+            )
+
+            error = cv2.norm(
+                src1=self.image_points[i],
+                src2=projected_image_points,
+                normType=cv2.NORM_L2,
+            )
+            error /= len(projected_image_points)
+            mean_error += error
+
+        self.logger.info(f"total eror: {mean_error / len(self.object_points)}")
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:  # type: ignore
```

Comment:
This method has quite a number of lines of code, possible to break up into sub-functions so that users can quickly tell at a glance what it does?

Reply:
Adding to the above - I think it is possible to cover more functions in unit testing and you may want to structure your functions partly for this purpose as well.
See how the returned value of a [function](https://github.com/aimakerspace/PeekingDuck/blob/f3144e0c5d39fb9b33ae4d79273e80256afb3ef7/peekingduck/pipeline/nodes/base.py#L323) was [mocked here](https://github.com/aimakerspace/PeekingDuck/blob/f3144e0c5d39fb9b33ae4d79273e80256afb3ef7/tests/pipeline/nodes/model/mtcnnv1/test_mtcnn.py#L92) for mtcnn's unit test.